src/main/java/net/neoforged/neoforge/client/entity/animation/json/AnimationParser.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.entity.animation.json;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.MapCodec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import it.unimi.dsi.fastutil.Pair;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.UnaryOperator;
+import net.minecraft.client.animation.AnimationChannel;
+import net.minecraft.client.animation.AnimationDefinition;
+import net.minecraft.client.animation.Keyframe;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.util.ExtraCodecs;
+import net.neoforged.neoforge.client.entity.animation.AnimationKeyframeTarget;
+import net.neoforged.neoforge.client.entity.animation.AnimationTarget;
+import org.joml.Vector3f;
+
+/**
+ * A parser for parsing JSON-based entity animation files.
+ */
+public final class AnimationParser {
+    private static final Codec<AnimationTarget> TARGET_CODEC = ResourceLocation.CODEC
+            .flatXmap(
+                    name -> Optional.ofNullable(AnimationTypeManager.getTarget(name))
+                            .map(DataResult::success)
+                            .orElseGet(() -> DataResult.error(() -> String.format(
+                                    Locale.ENGLISH, "Animation target '%s' not found. Registered targets: %s",
+                                    name, AnimationTypeManager.getTargetList()))),
+                    target -> Optional.ofNullable(AnimationTypeManager.getTargetName(target))
+                            .map(DataResult::success)
+                            .orElseGet(() -> DataResult.error(() -> String.format(
+                                    Locale.ENGLISH, "Unregistered animation target '%s'. Registered targets: %s",
+                                    target, AnimationTypeManager.getTargetList()))));
```

Comment:
This is just a gripe comment, but this can probably be put into a method since it's used multiple times. However, you can ignore me.

Reply:
I thought about doing this, but then I realized that this function would need a really large number of arguments such that it actually wouldn't end up deduplicating very much code.

Reply:
Fair enough.